File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/Native/AlxTopOnNativeObject.m
//
//  AlxTopOnNativeObject.m
//  AlxAdsOCDemo
//

#import "AlxTopOnNativeObject.h"
#import "AlxTopOnNativeEvent.h"

@implementation AlxTopOnNativeObject

#pragma mark - ATCustomNetworkNativeAd Override Methods

// ⚠️ 注意：TopOn SDK 会调用这个方法
- (BOOL)isExpressAd {
    // 自渲染广告返回 NO，模板广告返回 YES
    return NO;
}

// TopOn SDK 会从这些方法获取数据，而不是直接访问属性
- (NSString *)title {
    NSString *result = self.nativeAd.title ?: @"";
    NSLog(@"AlxTopOnNativeObject: title = %@", result);
    return result;
}

- (NSString *)mainText {
    NSString *result = self.nativeAd.desc ?: @"";
    NSLog(@"AlxTopOnNativeObject: mainText = %@", result);
    return result;
}

- (NSString *)iconUrl {
    NSString *result = self.nativeAd.icon.url ?: @"";
    NSLog(@"AlxTopOnNativeObject: iconUrl = %@", result);
    return result;
}

- (NSString *)mainImageUrl {
    NSString *result = @"";
    if (self.nativeAd.images.count > 0) {
        result = self.nativeAd.images.firstObject.url ?: @"";
    }
    NSLog(@"AlxTopOnNativeObject: mainImageUrl = %@", result);
    return result;
}

- (NSString *)ctaText {
    NSString *result = self.nativeAd.callToAction ?: @"";
    NSLog(@"AlxTopOnNativeObject: ctaText = %@", result);
    return result;
}

- (NSString *)advertiser {
    NSString *result = self.nativeAd.adSource ?: @"";
    NSLog(@"AlxTopOnNativeObject: advertiser = %@", result);
    return result;
}

- (UIImage *)logoImage {
    UIImage *result = self.nativeAd.adLogo;
    NSLog(@"AlxTopOnNativeObject: logoImage = %@", result ? @"exists" : @"nil");
    return result;
}

- (BOOL)isVideoContents {
    BOOL result = self.nativeAd.createType == 1;
    NSLog(@"AlxTopOnNativeObject: isVideoContents = %@", result ? @"YES" : @"NO");
    return result;
}

- (UIView *)mediaView {
    NSLog(@"AlxTopOnNativeObject: mediaView requested");
    
    // 如果已经创建过，直接返回缓存的实例，避免 TopOn 多次调用时返回不同对象
    if (self.cachedMediaView) {
        NSLog(@"AlxTopOnNativeObject: returning cached mediaView");
        return self.cachedMediaView;
    }
    
    // 否则创建新的 mediaView 并缓存
    if (self.nativeAd.mediaContent) {
        NSLog(@"AlxTopOnNativeObject: creating mediaView");
        AlxMediaView *mediaView = [[AlxMediaView alloc] init];
        [mediaView setMediaContent:self.nativeAd.mediaContent];
        self.cachedMediaView = mediaView;
        return mediaView;
    }
    
    NSLog(@"AlxTopOnNativeObject: no mediaContent");
    return nil;
}

- (CGFloat)mainImageWidth {
    CGFloat result = 0;
    if (self.nativeAd.images.count > 0) {
        result = self.nativeAd.images.firstObject.width;
    }
    NSLog(@"AlxTopOnNativeObject: mainImageWidth = %f", result);
    return result;
}

- (CGFloat)mainImageHeight {
    CGFloat result = 0;
    if (self.nativeAd.images.count > 0) {
        result = self.nativeAd.images.firstObject.height;
    }
    NSLog(@"AlxTopOnNativeObject: mainImageHeight = %f", result);
    return result;
}

#pragma mark - ATCustomNetworkNativeAd Required Methods

- (void)registerClickableViews:(NSArray<UIView *> *)clickableViews
                  withContainer:(UIView *)container
               registerArgument:(ATNativeRegisterArgument *)registerArgument {
    NSLog(@"AlxTopOnNativeObject: registerClickableViews");
    
    if (!self.nativeAd) {
        NSLog(@"AlxTopOnNativeObject: nativeAd is nil, cannot register");
        return;
    }
    
    if (container) {
        // ⚠️ 从 registerArgument 获取 viewController
        if (registerArgument.viewController) {
            self.nativeAd.rootViewController = registerArgument.viewController;
        }
        
        [self.nativeAd registerViewWithContainer:container
                                  clickableViews:clickableViews ?: @[container]];
    }
}

- (void)dealloc {
    NSLog(@"AlxTopOnNativeObject: dealloc");
    [self.nativeAd unregisterView];
    self.cachedMediaView = nil;
}

@end

File: custom_adapter/TopOnAdapter/Swift/6.5.x/Native/AlxTopOnNativeObject.swift
//
//  AlxTopOnNativeObject.swift
//  AlxAdsDemo
//
//  关键：父类属性是 readwrite，必须提供 setter / Key: superclass properties are readwrite, so setters must be provided
//

import Foundation
import AnyThinkSDK
import AlxAds

@objc(AlxTopOnNativeObject)
public class AlxTopOnNativeObject: ATCustomNetworkNativeAd {
    
    @objc public var nativeAd: AlxNativeAd?
    @objc public weak var nativeEvent: AlxTopOnNativeEvent?
    
    // ⚠️ 缓存 mediaView 实例，避免重复创建 / Cache the mediaView instance to avoid repeated creation
    private var _cachedMediaView: AlxMediaView?
    
    // MARK: - Override Readwrite Properties
    // ⚠️ 关键：父类是 readwrite，必须提供 getter 和 setter / Key: superclass is readwrite, must provide both getter and setter
    
    @objc public override var isExpressAd: Bool {
        get {
            return false
        }
        set {
            // 不需要实现，父类属性我们不修改 / No implementation needed, we don't modify the superclass property
        }
    }
    
    @objc public override var title: String? {
        get {
            let result = nativeAd?.title ?? ""
            NSLog("AlxTopOnNativeObject: title = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public override var mainText: String? {
        get {
            let result = nativeAd?.desc ?? ""
            NSLog("AlxTopOnNativeObject: mainText = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public override var iconUrl: String? {
        get {
            let result = nativeAd?.icon?.url ?? ""
            NSLog("AlxTopOnNativeObject: iconUrl = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public var mainImageUrl: String? {
        get {
            var result = ""
            if let images = nativeAd?.images, images.count > 0 {
                result = images.first?.url ?? ""
            }
            NSLog("AlxTopOnNativeObject: mainImageUrl = \(result)")
            return result
        }
    }
    
    @objc public override var ctaText: String? {
        get {
            let result = nativeAd?.callToAction ?? ""
            NSLog("AlxTopOnNativeObject: ctaText = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public override var advertiser: String? {
        get {
            let result = nativeAd?.adSource ?? ""
            NSLog("AlxTopOnNativeObject: advertiser = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public var logoImage: UIImage? {
        get {
            let result = nativeAd?.adLogo
            NSLog("AlxTopOnNativeObject: logoImage = \(result != nil ? "exists" : "nil")")
            return result
        }
    }
    
    @objc public override var isVideoContents: Bool {
        get {
            let result = nativeAd?.createType == 1
            NSLog("AlxTopOnNativeObject: isVideoContents = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public override var mediaView: UIView? {
        get {
            NSLog("AlxTopOnNativeObject: mediaView requested")
            
            // 如果已经创建过，直接返回缓存的实例 / If already created, return the cached instance directly
            if let cachedView = _cachedMediaView {
                NSLog("AlxTopOnNativeObject: returning cached mediaView")
                return cachedView
            }
            
            // 否则创建新的 mediaView / Otherwise create a new mediaView
            if let mediaContent = nativeAd?.mediaContent {
                NSLog("AlxTopOnNativeObject: creating new mediaView")
                let mediaView = AlxMediaView()
                mediaView.setMediaContent(mediaContent)
                _cachedMediaView = mediaView
                return mediaView
            }
            
            NSLog("AlxTopOnNativeObject: no mediaContent")
            return nil
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public override var mainImageWidth: CGFloat {
        get {
            var result: CGFloat = 0
            if let images = nativeAd?.images, images.count > 0 {
                result = CGFloat(images.first?.width ?? 0)
            }
            NSLog("AlxTopOnNativeObject: mainImageWidth = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    @objc public override var mainImageHeight: CGFloat {
        get {
            var result: CGFloat = 0
            if let images = nativeAd?.images, images.count > 0 {
                result = CGFloat(images.first?.height ?? 0)
            }
            NSLog("AlxTopOnNativeObject: mainImageHeight = \(result)")
            return result
        }
        set {
            // 不需要实现 / No implementation needed
        }
    }
    
    // MARK: - Override Method
    
    @objc public override func registerClickableViews(_ clickableViews: [UIView], 
                                                     withContainer container: UIView, 
                                                     registerArgument: ATNativeRegisterArgument?) {
        NSLog("AlxTopOnNativeObject: registerClickableViews")
        
        guard let nativeAd = self.nativeAd else {
            NSLog("AlxTopOnNativeObject: nativeAd is nil, cannot register")
            return
        }
        
        if let viewController = registerArgument?.viewController {
            nativeAd.rootViewController = viewController
        }
        
        let views = clickableViews.isEmpty ? [container] : clickableViews
        nativeAd.registerView(container: container, clickableViews: views)
    }
    
    deinit {
        NSLog("AlxTopOnNativeObject: dealloc")
        nativeAd?.unregisterView()
        _cachedMediaView = nil
    }
}
